Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+Serialization.swift

```diff
@@ -0,0 +1,183 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2023 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See https://swift.org/LICENSE.txt for license information
+ See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+import SymbolKit
+
+// MARK: PathHierarchy
+
+extension PathHierarchy.FileRepresentation {
+    // This mapping closure exist so that we don't encode ResolvedIdentifier values into the file. They're an implementation detail and they are a not stable across executions.
+    
+    /// Encode a path hierarchy into a file representation.
+    ///
+    /// The caller can use `mapCreatedIdentifiers` when encoding and decoding path hierarchies to associate auxiliary data with a node in the hierarchy.
+    ///
+    /// - Parameters:
+    ///   - fileRepresentation: A path hierarchy to encode.
+    ///   - mapCreatedIdentifiers: A closure that the caller can use to map indices to resolved identifiers.
+    init(
+        _ pathHierarchy: PathHierarchy,
+        mapCreatedIdentifiers: (_ identifiers: [ResolvedIdentifier]) -> Void
+    ) {
+        let lookup = pathHierarchy.lookup
+        
+        // Map each identifier to a number which will be used as to reference other nodes in the file representation.
+        var identifierMap = [ResolvedIdentifier: Int]()
+        identifierMap.reserveCapacity(lookup.count)
+        for (index, identifier) in zip(0..., lookup.keys) {
+            identifierMap[identifier] = index
+        }
+        
+        let nodes = [Node](unsafeUninitializedCapacity: lookup.count) { buffer, initializedCount in
```

Comment:
Sorry – I thought I asked this on the first review but it looks like I didn't. Is there a need for the `unsafe` initializer here? Or should we consider initially landing without it and then adopting it if there's a noticeable performance gain?

(I'm not familiar with the risk/reward of this particular initializer – the `unsafe` identifier is just raising alarm bells for me.)

Reply:
This initializer is described in [SE-0245][1] and was added in Swift 5.1. It allows the caller to initialize an array of a known size and initialize the elements out of order by writing into the array's buffer. The alternatives involve extra allocations and extra copying of data.

Even if each node is small, the list is long (one node for each page, landmark, and anchor). This initialization is still probably just a rounding error in the scope of the overall build but this API was added for specifically this use case (initializing an arrays elements out of order).

[1]: https://github.com/apple/swift-evolution/blob/main/proposals/0245-array-uninitialized-initializer.md

Reply:
Got it. Since this is the exact intent of this API – I think it makes sense to use it here. Thank you for the explanation! It might be nice to add a comment calling that out since we don't have a lot of `unsafe` API usage in the codebase generally, but non-blocking.